finat/hdivcurl.py

```diff
@@ -0,0 +1,168 @@
+        return self._transform_evaluation(core_eval)
+
+    def point_evaluation(self, order, refcoords, entity=None):
+        core_eval = self.wrapped.point_evaluation(order, refcoords, entity)
+        return self._transform_evaluation(core_eval)
+
+
+class HDivElement(WrapperElementBase):
+    """H(div) wrapper element for tensor product elements."""
+
+    def __init__(self, wrapped):
+        assert isinstance(wrapped, TensorProductElement)
+        if any(fe.formdegree is None for fe in wrapped.factors):
+            raise ValueError("Form degree of subelement is None, cannot H(div)!")
+
+        formdegree = sum(fe.formdegree for fe in wrapped.factors)
+        if formdegree != wrapped.cell.get_spatial_dimension() - 1:
+            raise ValueError("H(div) requires (n-1)-form element!")
+
+        transform = select_hdiv_transformer(wrapped)
+        super(HDivElement, self).__init__(wrapped, transform)
+
+    @property
+    def formdegree(self):
+        return self.cell.get_spatial_dimension() - 1
+
+    @property
+    def mapping(self):
+        return "contravariant piola"
+
+
+class HCurlElement(WrapperElementBase):
+    """H(curl) wrapper element for tensor product elements."""
+
+    def __init__(self, wrapped):
+        assert isinstance(wrapped, TensorProductElement)
+        if any(fe.formdegree is None for fe in wrapped.factors):
+            raise ValueError("Form degree of subelement is None, cannot H(curl)!")
+
+        formdegree = sum(fe.formdegree for fe in wrapped.factors)
+        if formdegree != 1:
+            raise ValueError("H(curl) requires 1-form element!")
+
+        transform = select_hcurl_transformer(wrapped)
+        super(HCurlElement, self).__init__(wrapped, transform)
+
+    @property
+    def formdegree(self):
+        return 1
+
+    @property
+    def mapping(self):
+        return "covariant piola"
+
+
+def select_hdiv_transformer(element):
+    # Assume: something x interval
+    assert len(element.factors) == 2
+    assert element.factors[1].cell.get_shape() == LINE
+
+    ks = tuple(fe.formdegree for fe in element.factors)
+    if ks == (0, 1):
+        return lambda v: [gem.Product(gem.Literal(-1), v), gem.Zero()]
+    elif ks == (1, 0):
+        return lambda v: [gem.Zero(), v]
+    elif ks == (2, 0):
+        return lambda v: [gem.Zero(), gem.Zero(), v]
+    elif ks == (1, 1):
+        if element.mapping == "contravariant piola":
+            return lambda v: [gem.Indexed(v, (0,)),
+                              gem.Indexed(v, (1,)),
+                              gem.Zero()]
+        elif element.mapping == "covariant piola":
+            return lambda v: [gem.Indexed(v, (1,)),
+                              gem.Product(gem.Literal(-1), gem.Indexed(v, (0,))),
+                              gem.Zero()]
+        else:
+            assert False, "Unexpected original mapping!"
+    else:
+        assert False, "Unexpected form degree combination!"
```

Comment:
So this is padding a lower-dimensional vector into an appropriately-sized spatial_dimension vector for Hdiv (i.e. continuous normals).  Maybe a comment as to why you do x -> [-x, 0] rather than [+x, 0].  This is to do with requirements of quads?